pkg/controller/route/routecontroller.go

```diff
@@ -26,73 +26,147 @@ import (
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return rc.kubeClient.Core().Nodes().Watch(options)
+			},
+		},
+		&api.Node{},
+		controller.NoResyncPeriodFunc(),
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				status := NodeStatus{
+					node:   obj.(*api.Node),
+					status: statusAdd,
+				}
+				rc.queue.Add(status)
+			},
+			UpdateFunc: func(old, new interface{}) {
+				status := NodeStatus{
+					node:   new.(*api.Node),
+					status: statusUpdate,
+				}
+				rc.queue.Add(status)
+			},
+			DeleteFunc: func(obj interface{}) {
+				status := NodeStatus{
+					node:   obj.(*api.Node),
+					status: statusDelete,
+				}
+				rc.queue.Add(status)
+			},
+		},
+	)
+	rc.syncHandler = rc.reconcileNodeRoutes
+	return rc
+}
+
+func (rc *RouteController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	go rc.nodeController.Run(wait.NeverStop)
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(rc.worker, time.Second, stopCh)
 	}
+	<-stopCh
+	rc.queue.ShutDown()
 }
 
-func (rc *RouteController) Run(syncPeriod time.Duration) {
-	// TODO: If we do just the full Resync every 5 minutes (default value)
-	// that means that we may wait up to 5 minutes before even starting
-	// creating a route for it. This is bad.
-	// We should have a watch on node and if we observe a new node (with CIDR?)
-	// trigger reconciliation for that node.
-	go wait.NonSlidingUntil(func() {
-		if err := rc.reconcileNodeRoutes(); err != nil {
-			glog.Errorf("Couldn't reconcile node routes: %v", err)
+func (rc *RouteController) worker() {
+	workFunc := func() bool {
+		status, quit := rc.queue.Get()
+		if quit {
+			return true
+		}
+		defer rc.queue.Done(status)
+		err := rc.syncHandler(status.(NodeStatus))
+		if err != nil {
+			glog.Errorf("Error syncing route: %v", err)
 		}
-	}, syncPeriod, wait.NeverStop)
+		return false
+	}
+	for {
+		if quit := workFunc(); quit {
+			glog.Infof("route controller worker shutting down")
+			return
+		}
+	}
 }
 
-func (rc *RouteController) reconcileNodeRoutes() error {
+func (rc *RouteController) reconcileNodeRoutes(status NodeStatus) error {
 	routeList, err := rc.routes.ListRoutes(rc.clusterName)
 	if err != nil {
 		return fmt.Errorf("error listing routes: %v", err)
 	}
-	// TODO (cjcullen): use pkg/controller/framework.NewInformer to watch this
-	// and reduce the number of lists needed.
-	nodeList, err := rc.kubeClient.Core().Nodes().List(api.ListOptions{})
-	if err != nil {
-		return fmt.Errorf("error listing nodes: %v", err)
-	}
-	return rc.reconcile(nodeList.Items, routeList)
+	return rc.reconcile(status, routeList)
 }
 
-func (rc *RouteController) reconcile(nodes []api.Node, routes []*cloudprovider.Route) error {
-	// nodeCIDRs maps nodeName->nodeCIDR
-	nodeCIDRs := make(map[string]string)
-	// routeMap maps routeTargetInstance->route
+func (rc *RouteController) reconcile(status NodeStatus, routes []*cloudprovider.Route) error {
```

Comment:
I'll take a look, maybe early next week.  I don't recall seeing anything about a rateLimiter in the code though.  The old code would cycle through all nodes every 30 seconds.  I don't see how this code would be more intensive than the old cold.  reconcile works on 1 node at a time now instead of all nodes, and only does so when there is a node added/changed/deleted.

It's possible I missed the rateLimiter though.  I'll take a look.